feat: add liquidity to existing position informations

NeOMakinG opened the pull request:
## Description

Pretty straightforwards from an implementation side, the difficulty is on all the cases we have to cover depending on the already opened positions

For reference here are the cases: https://github.com/shapeshift/web/issues/6681#issuecomment-2204450317

Also, we need to think that we can have 2 positions opened for this current pair meaning that we need to display different values depending on the already existing positions

<!-- Please describe your changes -->

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #6681

## Risk
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing
- Test to open a new position on every side: asym for both assets and sym
- Then test to deposit using different side on already existing positions
- Then test to deposit on the same side of an already existing position and verify that the pool is updated as expected
- Test to only have a rune side position, sym should be disabled

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering
n/a
<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations
n/a
<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

## New position
![image](https://github.com/user-attachments/assets/b82d2b2d-48f0-44f4-8ec6-7f2dab99ae34)

## Asym rune
![image](https://github.com/user-attachments/assets/99801eca-6b60-492e-9d41-f939139dada3)

## Asym rune -> Asym Asset
![image](https://github.com/user-attachments/assets/75001135-34e4-480a-9b51-8cd152b74062)

## Asym Asset
![image](https://github.com/user-attachments/assets/8bf272d6-c3e1-44df-aa26-3c93835b76d6)

## Asym Asset => Sym deposit
![image](https://github.com/user-attachments/assets/10075d34-a755-4a51-a3e5-07fbe14f5d41)

## Sym
![image](https://github.com/user-attachments/assets/595e9346-c69e-4b9f-aa40-6ebce4e13111)

## Sym -> Asym Asset deposit
![image](https://github.com/user-attachments/assets/eb447476-18a1-4b2e-bd76-84f4fa2238e5)

0xApotheosis commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+    }
+
+    const runeInformations: PositionInformations = (() => {
+      if (hasSymPosition) return displayAmounts('sym')
```
This might actually be where the bug comes from where it is showing my asym balance in the RUNE sym spot (instead of new position).

I suspect we need to re-work this logic.

NeOMakinG replied:
No, the logic seems good, if you have a `sym` position and tries to add some rune, you will simply add it to the sym position, you can't open a rune position if you already have a `sym` position opened

We might be able to find a clearer way to display this but currently it's not a bug but a feature!

0xApotheosis commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+      return newPosition
+    })()
+
+    const assetInformations: PositionInformations = (() => {
+      if (hasAsymRunePosition || hasSymPosition) {
+        return hasAsymAssetPosition ? displayAmounts(AsymSide.Asset) : newPosition
+      }
+      return hasAsymAssetPosition ? displayAmounts(AsymSide.Asset) : newPosition
+    })()
```
This is confusing, it seems that the if statement is redundant. I think the below would do the same thing:

```suggestion
    const assetInformations: PositionInformations = (() => {
      return hasAsymAssetPosition ? displayAmounts(AsymSide.Asset) : newPosition
    })()
```

NeOMakinG replied:
You're right, changed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

0xApotheosis commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+      return hasAsymAssetPosition ? displayAmounts(AsymSide.Asset) : newPosition
+    })()
+
+    const symInformations: PositionInformations = (() => {
+      if (hasAsymRunePosition) return notSupported
+      if (hasAsymAssetPosition) {
+        return hasSymPosition ? displayAmounts('sym') : newPosition
+      }
+      if (hasSymPosition) return displayAmounts('sym')
+
+      return newPosition
+    })()
```
Also quite confusing. Perhaps:

```suggestion
    const symInformations: PositionInformations = (() => {
      if (hasAsymRunePosition) return notSupported
      if (hasSymPosition) return displayAmounts('sym')
      return newPosition
    })()
```

NeOMakinG replied:
Changed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

0xApotheosis commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+      tooltipText: translate('pools.notSupportedTooltip'),
+    }
+
+    const runeInformations: PositionInformations = (() => {
+      if (hasSymPosition) return displayAmounts('sym')
+      if (hasAsymAssetPosition) {
+        return hasAsymRunePosition ? displayAmounts(AsymSide.Rune) : newPosition
+      }
+      if (hasAsymRunePosition) return displayAmounts(AsymSide.Rune)
+
+      return newPosition
+    })()
```
The below is a simpler version of the current logic:

```suggestion
    const runeInformations: PositionInformations = (() => {
      if (hasSymPosition) return displayAmounts('sym')
      if (hasAsymRunePosition) return displayAmounts(AsymSide.Rune)
      return newPosition
    })()
```

However, it's not clear to me why we don't show return `newPosition` if `!hasAsymRunePosition`?

Same with the logic below.

NeOMakinG replied:
With you proposition, we actually show return `newPosition` if `!hasAsymRunePosition`

If we already have a symetrical position, using the `rune` deposit will effectively deposit to the `sym` pool instead of depositing to the asym rune pool

Changed to your implementation in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

0xApotheosis reviewed (changes requested):
First pass testing below.

I think we need to re-work the logic inside `informationsByPosition` (and give it a more descriptive name), as it's quite confusing (and possibly broken) as it is currently.

✅ New position:

<img width="325" alt="Screenshot 2024-07-29 at 12 26 57 PM" src="https://github.com/user-attachments/assets/aa6ae822-81e6-4b46-bca2-043d8db656cd">

✅ Existing asym:

<img width="321" alt="Screenshot 2024-07-29 at 12 25 50 PM" src="https://github.com/user-attachments/assets/b49f1852-c862-42b7-98c4-8960ea9c908b">

✅  Existing asset asym position (DOGE):

<img width="314" alt="Screenshot 2024-07-29 at 12 24 10 PM" src="https://github.com/user-attachments/assets/1f70d226-158c-4a15-a1ce-68d0246e2744">

✅ Existing sym and asym RUNE:

<img width="316" alt="Screenshot 2024-07-29 at 12 28 10 PM" src="https://github.com/user-attachments/assets/02d98d95-4a0a-4514-8799-255366464ffb">

✅ Existing asym

<img width="310" alt="Screenshot 2024-07-29 at 12 39 45 PM" src="https://github.com/user-attachments/assets/11f86d35-cab9-42df-87ba-6413f4c7ed80">

❌ Position display (blocking):

For AVAX, I have an Asymmetric AVAX position and a symmetric position. I have no RUNE asym position.

It seems off that the RUNE position is showing a balance, too (this same as the asym balance)?

<img width="310" alt="Screenshot 2024-07-29 at 12 57 41 PM" src="https://github.com/user-attachments/assets/56f01234-7c39-4a8f-8bfc-4c2a55130fab">

NeOMakinG commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -157,8 +292,15 @@ export const LpType = ({
+            label={translate(
+              isDeposit && currentSideInformations.tooltipText
+                ? currentSideInformations.tooltipText
+                : 'pools.symWithdrawOnAsymPositionAlert',
```
Damn, wrong mergefix after all the reviews feedbacks of the initial PR, changed it to the initial translation!

There was a problem with the Tooltip `isDisabled` prop condition too!

Fixed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

NeOMakinG commented on src/assets/translations/en/main.json:
```diff
@@ -2356,6 +2356,10 @@
     "removeAmounts": "Remove Amount(s)",
     "stagedTooltip": "Staged pools are new pools accepting liquidity, but will not accrue any earnings until Available",
+    "newPosition": "New position",
+    "newPositionTooltip": "This will create a new LP position instead of adding to the existing one.",
```
Changed it to `any existing ones` in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

NeOMakinG commented on src/pages/ThorChainLP/components/AddLiquidity/AddLiquidityInput.tsx:
```diff
@@ -338,11 +339,46 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
+          underlyingRuneAmountCryptoPrecision: position.underlyingRuneAmountCryptoPrecision,
+        },
+      }
+    }, {} as AmountsByPosition)
```
Sweet, fixed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

NeOMakinG commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+    }
+  }, [])
+
+  const informationsByPosition = useMemo(() => {
```
Fixed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

NeOMakinG commented on src/pages/ThorChainLP/components/LpType.tsx:
```diff
@@ -139,12 +181,105 @@ export const LpType = ({
+      tooltipText: translate('pools.notSupportedTooltip'),
+    }
+
+    const runeInformations: PositionInformations = (() => {
```
Changed in [e22aec7](https://github.com/shapeshift/web/pull/7367/commits/e22aec780b3128ece64322e6a640b9af9d997090)

NeOMakinG commented:
@0xApotheosis thanks for the high quality review as usually! I addressed your feedbacks.

If you have a `sym` position, using the `asym rune` side to deposit will add some liquidity to the `sym` pool instead of creating one, this is why we are displaying the amounts of the sym pool, might be a product thing but this is how it works and how I thought it, so it's working as expected.

Unless the product team wants us to display something else, it shouldn't be a blocker and reflects the way it works for now on the THORChain side!

0xApotheosis reviewed (approved):
Review feedback addressed, thank you ser - looking good.
